# EDA Notebook
Voy a hacer una primera revisión del dataset para averiguar que debo limpiar. Tras la limpieza, haré una nueva revisión para observar cómo son los datos (ya limpios).

## 1. Importar librerías

In [72]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import tqdm                    # A priori se utilizará para ver el progreso de algunos procesos, que con tantos datos puede tardar.

## 2. Set Up

### Condiciones del display
El objetivo es que todos los display se vean iguales y obligarlo a mostrar todas las columnas para poder visualizar mejor los datos.

In [73]:
# Mostrar todas las columnas (sin cortar)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

# Mostrar decimales con 2 cifras nada más.
pd.set_option('display.float_format', '{:.2f}'.format)

### Cargar el dataset
No sólo voy a cargar el archivo, sino que voy a eliminar directamente algunas columnas que sé que no voy a utilizar debido a las preguntas que quiero responder. De esta manera, reduzco el volumen de datos, que con un dataset tan grande es importante.

Las columnas eliminadas de primeras son: State, MeetState, MeetTown, MeetCountry y ParentFederation.

Tras la limpieza, eliminaré más columnas, pero ahora, las necesito para imputar los datos y filtrar.

In [74]:
opl_df = pd.read_csv('../data/OPL_dataset.csv')

raw_df = opl_df.drop(columns = ['State', 'MeetState', 'MeetTown', 'MeetCountry', 'ParentFederation', 'MeetName', 'WeightClassKg', 'Division'])

C:\Users\Lap-user\AppData\Local\Temp\ipykernel_2456\1402644969.py:1: DtypeWarning: Columns (0: State, 1: ParentFederation, 2: MeetState) have mixed types. Specify dtype option on import or set low_memory=False.
  opl_df = pd.read_csv('../data/OPL_dataset.csv')


## 3. Explorar el dataset

### 3.1. Exploración general

In [75]:
display(raw_df.head())

,Name,Sex,Event,Equipment,Age,AgeClass,BirthYearClass,BodyweightKg,Squat1Kg,Squat2Kg,Squat3Kg,Squat4Kg,Best3SquatKg,Bench1Kg,Bench2Kg,Bench3Kg,Bench4Kg,Best3BenchKg,Deadlift1Kg,Deadlift2Kg,Deadlift3Kg,Deadlift4Kg,Best3DeadliftKg,TotalKg,Place,Dots,Wilks,Glossbrenner,Goodlift,Tested,Country,Federation,Date,Sanctioned
0,E.S. Denisenko,F,B,Raw,28.50,24-34,24-39,67.30,NaN,NaN,NaN,NaN,NaN,-40.00,-45.00,-45.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,DQ,NaN,NaN,NaN,NaN,NaN,NaN,GSF-Belarus,2019-06-22,Yes
1,I.S. Lebetskaya,F,B,Raw,43.50,40-44,40-49,73.20,NaN,NaN,NaN,NaN,NaN,80.00,85.00,90.00,NaN,90.00,NaN,NaN,NaN,NaN,NaN,90.00,1,88.80,86.89,76.50,70.06,NaN,NaN,GSF-Belarus,2019-06-22,Yes
2,K. Yakimovich,F,B,Raw,26.50,24-34,24-39,60.60,NaN,NaN,NaN,NaN,NaN,40.00,42.50,45.00,NaN,45.00,NaN,NaN,NaN,NaN,NaN,45.00,2,49.57,49.79,43.98,38.42,NaN,NaN,GSF-Belarus,2019-06-22,Yes
3,A.G. Golneva,F,B,Raw,19.50,20-23,19-23,50.30,NaN,NaN,NaN,NaN,NaN,32.50,35.00,-37.50,NaN,35.00,NaN,NaN,NaN,NaN,NaN,35.00,2,43.67,44.76,39.73,34.55,NaN,NaN,GSF-Belarus,2019-06-22,Yes
4,E.V. Marunevskaya,F,B,Raw,19.50,20-23,19-23,63.70,NaN,NaN,NaN,NaN,NaN,40.00,42.50,-45.00,NaN,42.50,NaN,NaN,NaN,NaN,NaN,42.50,1,45.40,45.26,39.96,35.25,NaN,NaN,GSF-Belarus,2019-06-22,Yes


### 3.2. Filas, columnas y tipos

In [76]:
raw_df.info()
print("---------------------------------------")
print(f"Número de filas: {raw_df.shape[0]}")
print(f"Número de columnas: {raw_df.shape[1]}")

<class 'pandas.DataFrame'>
RangeIndex: 3920662 entries, 0 to 3920661
Data columns (total 34 columns):
 #   Column           Dtype  
---  ------           -----  
 0   Name             str    
 1   Sex              str    
 2   Event            str    
 3   Equipment        str    
 4   Age              float64
 5   AgeClass         str    
 6   BirthYearClass   str    
 7   BodyweightKg     float64
 8   Squat1Kg         float64
 9   Squat2Kg         float64
 10  Squat3Kg         float64
 11  Squat4Kg         float64
 12  Best3SquatKg     float64
 13  Bench1Kg         float64
 14  Bench2Kg         float64
 15  Bench3Kg         float64
 16  Bench4Kg         float64
 17  Best3BenchKg     float64
 18  Deadlift1Kg      float64
 19  Deadlift2Kg      float64
 20  Deadlift3Kg      float64
 21  Deadlift4Kg      float64
 22  Best3DeadliftKg  float64
 23  TotalKg          float64
 24  Place            str    
 25  Dots             float64
 26  Wilks            float64
 27  Glossbrenner     float6

### 3.3. Estadística de las variables

In [77]:
print("----------------- ANÁLISIS VARIABLES NUMÉRICAS -----------------")
display(raw_df.describe())


print("----------------- ANÁLISIS VARIABLES CATEGÓRICAS -----------------")
print(" 1. Sexo")
display(raw_df['Sex'].value_counts())

print(" 2. Levantamientos realizados")
display(raw_df['Event'].value_counts())

print(" 3. Equipamiento permitido")
display(raw_df['Equipment'].value_counts())

print(" 4. Podium")
display(raw_df['Place'].value_counts())

print(" 5. Federación")
display(raw_df['Federation'].value_counts())

print(" 6. Competición 'Naty'")
display(raw_df['Tested'].value_counts())

print(" 7. Competición 'Real'")
display(raw_df['Sanctioned'].value_counts())

print(" 8. País")
display(raw_df['Country'].value_counts())

----------------- ANÁLISIS VARIABLES NUMÉRICAS -----------------


,Age,BodyweightKg,Squat1Kg,Squat2Kg,Squat3Kg,Squat4Kg,Best3SquatKg,Bench1Kg,Bench2Kg,Bench3Kg,Bench4Kg,Best3BenchKg,Deadlift1Kg,Deadlift2Kg,Deadlift3Kg,Deadlift4Kg,Best3DeadliftKg,TotalKg,Dots,Wilks,Glossbrenner,Goodlift
count,2470198.00,3876424.00,1260845.00,1247917.00,1217438.00,9733.00,2740168.00,1830613.00,1809555.00,1756605.00,24041.00,3477139.00,1410644.00,1386656.00,1336712.00,27359.00,2939414.00,3654734.00,3628632.00,3628632.00,3628632.00,3328872.00
mean,30.64,83.75,117.75,103.50,45.05,74.97,170.91,83.98,59.58,-11.92,23.76,114.56,160.41,136.75,26.20,77.00,186.39,380.48,280.38,279.23,262.36,64.33
std,13.31,22.71,132.56,156.20,188.13,178.77,67.13,93.73,118.42,135.82,158.60,52.61,105.81,152.17,210.50,186.83,62.21,200.59,125.47,124.93,118.72,16.43
min,0.00,15.00,-555.00,-600.01,-600.50,-550.00,-508.02,-635.50,-635.50,-589.99,-589.99,-522.50,-500.00,-502.50,-587.50,-500.00,-410.00,1.00,0.68,0.67,0.64,0.50
25%,20.50,66.80,92.50,85.00,-145.50,-90.00,120.00,57.50,37.50,-130.00,-116.00,72.50,125.00,120.00,-200.00,-106.80,137.50,212.50,164.54,163.93,153.82,52.39
50%,27.00,81.60,147.50,150.00,120.00,130.00,165.56,100.00,95.00,-55.00,73.00,111.10,180.00,180.00,127.50,142.50,185.00,360.00,302.30,301.19,280.25,64.16
75%,38.00,98.30,197.50,202.50,195.00,192.78,215.00,140.00,140.00,115.00,149.69,147.50,222.50,230.00,207.50,205.00,230.00,532.50,373.55,371.84,351.30,75.81
max,105.50,300.00,560.01,577.50,595.00,592.39,595.00,659.00,521.63,635.50,567.44,659.00,458.00,470.00,487.50,520.00,487.50,1407.50,818.06,813.18,756.90,182.71


----------------- ANÁLISIS VARIABLES CATEGÓRICAS -----------------
 1. Sexo


Sex
M     2825647
F     1094849
Mx        166
Name: count, dtype: int64

 2. Levantamientos realizados


Event
SBD    2842395
B       767193
D       211430
BD       73680
S        19473
SB        4157
SD        2334
Name: count, dtype: int64

 3. Equipamiento permitido


Equipment
Raw           1843211
Single-ply    1387396
Unlimited      323461
Wraps          236120
Multi-ply      130355
Straps            119
Name: count, dtype: int64

 4. Podium


Place
1      1612796
2       608648
3       369711
4       249769
DQ      241997
        ...   
150          1
151          1
152          1
195          1
124          1
Name: count, Length: 156, dtype: int64

 5. Federación


Federation
THSPA         535620
USAPL         359073
THSWPA        325029
USPA          226478
FPR           191693
               ...  
UgandaPA          12
USABPA            10
AmericanSA         9
DBKV               9
NOVA               7
Name: count, Length: 465, dtype: int64

 6. Competición 'Naty'


Tested
Yes    2984454
Name: count, dtype: int64

 7. Competición 'Real'


Sanctioned
Yes    3910195
No       10467
Name: count, dtype: int64

 8. País


Country
USA              1046973
Russia            279698
Canada             88269
Ukraine            86999
England            71801
                  ...   
Comoros                1
Lesotho                1
Guinea-Bissau          1
Djibouti               1
Mauritius              1
Name: count, Length: 195, dtype: int64

### 3.4. Nulos y duplicados

In [78]:
print(raw_df.isnull().sum())
print("--------------------------------------------")
print(f"Nº duplicados: {raw_df.duplicated().sum()}")

Name                     0
Sex                      0
Event                    0
Equipment                0
Age                1450464
AgeClass           1081168
BirthYearClass     1366527
BodyweightKg         44238
Squat1Kg           2659817
Squat2Kg           2672745
Squat3Kg           2703224
Squat4Kg           3910929
Best3SquatKg       1180494
Bench1Kg           2090049
Bench2Kg           2111107
Bench3Kg           2164057
Bench4Kg           3896621
Best3BenchKg        443523
Deadlift1Kg        2510018
Deadlift2Kg        2534006
Deadlift3Kg        2583950
Deadlift4Kg        3893303
Best3DeadliftKg     981248
TotalKg             265928
Place                    0
Dots                292030
Wilks               292030
Glossbrenner        292030
Goodlift            591790
Tested              936208
Country            1729709
Federation               0
Date                     0
Sanctioned               0
dtype: int64
--------------------------------------------
Nº duplicados: 93053


# Limpieza de los datos

## Gestión de duplicados
El análisis anterior muestra 93053 duplicados exactos → Eliminarlos [manteniendo uno (cualquiera)].

In [79]:
dp_df = raw_df.drop_duplicates()
display(dp_df.head())

,Name,Sex,Event,Equipment,Age,AgeClass,BirthYearClass,BodyweightKg,Squat1Kg,Squat2Kg,Squat3Kg,Squat4Kg,Best3SquatKg,Bench1Kg,Bench2Kg,Bench3Kg,Bench4Kg,Best3BenchKg,Deadlift1Kg,Deadlift2Kg,Deadlift3Kg,Deadlift4Kg,Best3DeadliftKg,TotalKg,Place,Dots,Wilks,Glossbrenner,Goodlift,Tested,Country,Federation,Date,Sanctioned
0,E.S. Denisenko,F,B,Raw,28.50,24-34,24-39,67.30,NaN,NaN,NaN,NaN,NaN,-40.00,-45.00,-45.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,DQ,NaN,NaN,NaN,NaN,NaN,NaN,GSF-Belarus,2019-06-22,Yes
1,I.S. Lebetskaya,F,B,Raw,43.50,40-44,40-49,73.20,NaN,NaN,NaN,NaN,NaN,80.00,85.00,90.00,NaN,90.00,NaN,NaN,NaN,NaN,NaN,90.00,1,88.80,86.89,76.50,70.06,NaN,NaN,GSF-Belarus,2019-06-22,Yes
2,K. Yakimovich,F,B,Raw,26.50,24-34,24-39,60.60,NaN,NaN,NaN,NaN,NaN,40.00,42.50,45.00,NaN,45.00,NaN,NaN,NaN,NaN,NaN,45.00,2,49.57,49.79,43.98,38.42,NaN,NaN,GSF-Belarus,2019-06-22,Yes
3,A.G. Golneva,F,B,Raw,19.50,20-23,19-23,50.30,NaN,NaN,NaN,NaN,NaN,32.50,35.00,-37.50,NaN,35.00,NaN,NaN,NaN,NaN,NaN,35.00,2,43.67,44.76,39.73,34.55,NaN,NaN,GSF-Belarus,2019-06-22,Yes
4,E.V. Marunevskaya,F,B,Raw,19.50,20-23,19-23,63.70,NaN,NaN,NaN,NaN,NaN,40.00,42.50,-45.00,NaN,42.50,NaN,NaN,NaN,NaN,NaN,42.50,1,45.40,45.26,39.96,35.25,NaN,NaN,GSF-Belarus,2019-06-22,Yes


## Gestión de tipos

- Columnas con tipo correcto[tipo actual]: Name `[Str]`, Age `[Float]`, Bodyweight `[Float]`, BenchANY `[Float]`, SquadANY `[Float]`, DeadliftANY `[Float]`, Total `[Float]`, ValoracionesANY `[Float]`.
- Columnas con tipo incorrecto[tipo esperado]: Sex `[Cat]`, Event `[Cat]`, Equipment `[Cat]`, Tested `[Boolean]`, Date `[Date]`, Sanctioned `[Bool]`, Country `[Cat]`, Federation `[Cat]`, Place `[Cat]`.

### Justificación de algunas decisiones
`Country`, `Federation` y `Place` son columnas con +100 posibles valores, lo que me generaba dudas sobre si debería convertirlas o no, pero como el objetivo es reducir espacio de la memoria, simplemente hice los cálculos y ví que el ahorro era tremendo. Consecuentemente, mantuve los cambios.

PD: Algunas columnas no han sufrido modificaciones (a pesar de que las necesitan) porque las mantengo en el dataset para que me ayuden a imputar los datos o a filtrar. Un ejemplo de ello es `AgeClass` o `BrithYearClass`

### Cambios

In [ ]:
# Categorías
types_df = dp_df.copy()
cat_var = ['Sex', 'Event', 'Equipment', 'Country', 'Federation', 'Place']
types_df[cat_var] = types_df[cat_var].astype('category')

# Fechas
types_df['Date'] = pd.to_datetime(types_df['Date'])

# Booleanas
types_df['Sanctioned'] = types_df['Sanctioned'].astype('bool')                      # BOOL porque no tiene NaN

types_df['Tested'] = types_df['Tested'].map({'Yes': True}).astype('boolean')        # Sin esto no puede convertir a BOOLEAN
types_df['Tested'] = types_df['Tested'].astype('boolean')                           # BOOLEAN porque tiene NaN

### Comprobaciones

In [ ]:
print("------------ AHORRO GENERAL ------------")
print(f" Memoria PRE-CLEAN: {dp_df.memory_usage(deep=True).sum()}")
print(f" Memoria POST-CLEAN: {types_df.memory_usage(deep=True).sum()}")

print("\n------------ AHORRO FEDERACIÓN ------------")
print(f"Federación ANTES  : {dp_df['Federation'].memory_usage(deep=True)}")
print(f"Federación DESPUÉS: {types_df['Federation'].memory_usage(deep=True)}")

print("\n------------ AHORRO PAÍS ------------")
print(f"País ANTES  : {dp_df['Country'].memory_usage(deep=True)}")
print(f"País DESPUÉS: {types_df['Country'].memory_usage(deep=True)}")

print("\n------------ AHORRO PODIUM ------------")
print(f"Podium ANTES  : {dp_df['Place'].memory_usage(deep=True)}")
print(f"Podium DESPUÉS: {types_df['Place'].memory_usage(deep=True)}")


print("\n------------ CHECK CAMBIOS ------------")
display(types_df.dtypes)


------------ AHORRO GENERAL ------------
 Memoria PRE-CLEAN: 3085114933
 Memoria POST-CLEAN: 1389806850

------------ AHORRO FEDERACIÓN ------------
Federation ANTES  : 235706098
Federation DESPUÉS: 38301803

------------ AHORRO PAÍS ------------
País ANTES  : 199542671
País DESPUÉS: 38287242

------------ AHORRO PODIUM ------------
Podium ANTES  : 222499298
Podium DESPUÉS: 38284089

------------ CHECK CAMBIOS ------------


Name                          str
Sex                      category
Event                    category
Equipment                category
Age                       float64
AgeClass                      str
BirthYearClass                str
BodyweightKg              float64
Squat1Kg                  float64
Squat2Kg                  float64
Squat3Kg                  float64
Squat4Kg                  float64
Best3SquatKg              float64
Bench1Kg                  float64
Bench2Kg                  float64
Bench3Kg                  float64
Bench4Kg                  float64
Best3BenchKg              float64
Deadlift1Kg               float64
Deadlift2Kg               float64
Deadlift3Kg               float64
Deadlift4Kg               float64
Best3DeadliftKg           float64
TotalKg                   float64
Place                    category
Dots                      float64
Wilks                     float64
Glossbrenner              float64
Goodlift                  float64
Tested        

### Análisis univariado de variables numéricas
Realizar histogramas y boxplots para analizar la distribución de las variables numéricas.

In [81]:
# Histogram for a numerical variable
# df['numeric_column'].hist()
# plt.show()

# Boxplot
# df.boxplot(column='numeric_column')
# plt.show()

### Análisis univariado de variables categóricas
Realizar gráficos de barras para analizar la frecuencia de las variables categóricas.

In [82]:
# Bar plot for a categorical variable
# df['categorical_column'].value_counts().plot(kind='bar')
# plt.show()

### Análisis bivariado entre variables
Explorar relaciones entre variables numéricas y categóricas mediante scatterplots y tablas cruzadas.

In [83]:
# Scatterplot
# plt.scatter(df['numeric_column_x'], df['numeric_column_y'])
# plt.show()

# Crosstab for two categorical variables   
# pd.crosstab(df['categorical_column1'], df['categorical_column2'])

### Visualización de correlaciones
Calcular y visualizar la matriz de correlación entre variables numéricas usando un heatmap.

In [84]:
# Correlation matrix
corr = opl_df.corr()
sns.heatmap(corr, annot=True)
plt.show()

ValueError: could not convert string to float: 'E.S. Denisenko'